docs/reST/ref/mouse.rst

```diff
@@ -183,77 +183,33 @@ scroll, such as ``which`` (it will tell you what exact mouse device trigger the
 
 .. function:: set_cursor
 
-   | :sl:`set the image for the mouse cursor`
-   | :sg:`set_cursor(size, hotspot, xormasks, andmasks) -> None`
+   | :sl:`set the mouse cursor to a new cursor`
+   | :sg:`set_cursor(pygame.cursors.Cursor) -> None`
+   | :sg:`set_cursor(arguments to create a pygame.cursors.Cursor) -> None`
```

Comment:
I guess this:
```
    | :sg:`set_cursor(pygame.cursors.Cursor) -> None`
    | :sg:`set_cursor(arguments to create a pygame.cursors.Cursor) -> None`
```

Could be replaced by:
```
   | :sg:`set_cursor(pygame.cursors.Cursor) -> None`
   | :sg:`set_cursor(size, hotspot, xormasks, andmasks) -> Cursor`
   | :sg:`set_cursor(hotspot, surface) -> Cursor`
   | :sg:`set_cursor(constant) -> Cursor`
```

However, `set_cursor(constant)` isn't really clear here in this list what should be used. Perhaps there is a way to make that more clear.

Have you considered python kwargs instead?

```python
set_cursor(arrow=True)
set_cursor(ibeam=True)
set_cursor(wait=True)
set_cursor(crosshair=True)
set_cursor(waitarrow=True)
set_cursor(size_nwse=True)
set_cursor(size_we=True)
set_cursor(size_ns=True)
set_cursor(size=True)
set_cursor(no=True)
set_cursor(hand=True)
```

Reply:
I think that many key word arguments would make for a very large function definition in python. The sort of thing pylint would normally cry about I would think. Though maybe it doesn't matter so much in functions defined in C, or kword only arguments. I don't think I've ever seen a function with 11 key word only arguments though. 

Could maybe just change it to:

```
| :sg:`set_cursor(cursor_constant) -> Cursor`
```

So it's more clear what 'constant' means, then there is a list of cursor constants right below theses signatures in the docs so people should be able to connect the two.

Reply:
The benefit of kwargs:
- they can be documented (so mouse hover on the arg shows the docs for example)
- they work with command line completion
- they are listed in the function and not off in a separate namespace that you need to read the docs to reference.
- constants are more of a C API thing, and not very common in python code.
- NOT_SHOUTING is easier to type 
- easier to read `set_cursor(arrow=True)` VS `set_cursor(pygame.SYSTEM_CURSOR_ARROW))`

We chose the kwarg approach for the Window object after much debate and experimentation.

The other option is set_cursor_arrow() et al. However this makes a proliferation of methods.

Reply:
Another approach would be:
```
mouse.cursor.arrow = True
```

Reply:
Looks like the default pylint warning for too many arguments is at 6, if that is a concern for the linting plans in the future.

Also for the kwargs approach what happens if someone does:

```
set_cursor(arrow=True, ibeam=True, wait=True)
```
?

The Enum type would be perfect for this use case (a limited but large-ish number of options) I think, but sadly it's not in python 2.7 by default so is probably out for pygame:

![image](https://user-images.githubusercontent.com/13382426/99907651-e9607100-2cd5-11eb-8ac4-fa79467370fe.png)

Code:

```
from enum import Enum


class SystemCursor(Enum):
    hand = 0
    arrow = 1
    wait = 2
    crosshair = 3


def set_cursor(system_cursor: SystemCursor):
    print(system_cursor)


set_cursor(SystemCursor.arrow)
```

Reply:
> Looks like the default pylint warning for too many arguments is at 6, if that is a concern for the linting plans in the future.

It's not a worry, since adding a `#pylint: disable=too-many-arguments` annotation can be done.

pylint doesn't have reasoning for this rule... However, eslint gives a good description:

> "Functions that take numerous parameters can be difficult to read and write because it requires the memorization of what each parameter is, its type, and the order they should appear in. As a result, many coders adhere to a convention that caps the number of parameters a function can take."

This guideline doesn't apply for this function, since you only need to use one of them. Kwargs in python don't require a fixed order, or require all the arguments to be used. X11 famously has lots of arguments for its functions. However, they don't have default arguments, and you have to at the call site use all the arguments. The solution in C of limiting the arguments and putting them into flags is good for C, flags in C let you do something like kwargs.

However, In this case, people only call it with one argument since python has kwarg defaults. So the pylint too-many-arguments does not apply here. No memorization of ordering needed.

```python
set_cursor(arrow=True)
```

> Also for the kwargs approach what happens if someone does:
> set_cursor(arrow=True, ibeam=True, wait=True)

It's not likely that someone would do that, but if they do then it raises an error telling them to use only one option.

> set_cursor(SystemCursor.arrow)

I think this is similar to the constants - you need to find where SystemCursor is and how to use it.

TypeScript allows you to use simple string union/enums like 'arrow' | 'ibeam' | 'wait', but I'm not sure this has made it into python world yet.

Reply:
The idea of the cursor object is to standardize the now 3 different ways of making a cursor into one API.

Passing in arguments to create a system cursor, or any type of cursor, is like how you can pass a sequence of numbers to a method expecting a rect. Set_cursor accepts a cursor or a "cursor like object", just like a lot of the draw functions accept a rect or a "rect like object."

I don't think that separating one type of cursors into its own thing ( keyword arguments ) is a good step forward.

I suppose each type could have its own keyword argument, but having only keyword arguments would break backwards compatibility on `set_cursor'. So having 0 kwargs would default to bitmap cursor, but then color cursor and system cursor would need kwargs?

Using a constant from pygame is something you have to do in lots of other places, and I think it is the simplest way to keep each type of cursor on the same playing field.

Reply:
> I don't think that separating one type of cursors into its own thing ( keyword arguments ) is a good step forward.

Why?

Reply:
Because I think having a different style for one type than for the others increases complexity rather than reducing it.

    pygame.mouse.set_cursor((24,24), (8,8), xormasks, andmasks)
    pygame.mouse.set_cursor((5,10), pygame.Surface((20,20)))
    pygame.mouse.set_cursor(pygame.SYSTEM_CURSOR_WAIT)

    pygame.mouse.set_cursor((24,24), (8,8), xormasks, andmasks)
    pygame.mouse.set_cursor((5,10), pygame.Surface((20,20)))
    pygame.mouse.set_cursor(wait=True) #although this does look prettier

Also I think kwargs should be for additional information or arguments, for stuff you can mix and match, not for the entire substance of a call. Having a pattern where you can call one kwarg OR two different function signatures seems weird to me.

Anyways I have to go to sleep now, talk to you later.

Reply:
[Literal types](https://www.python.org/dev/peps/pep-0586/#abstract) let you specify valid inputs, which provides an elegant interface...

```python
SystemCursors = Literal["arrow", "ibeam", "wait", "crosshair", "waitarrow",
    "size_nwse", "size_we", "size_ns", "size", "no", "hand"]
def set_cursor(cursor: SystemCursors) -> None:
    pass

set_cursor("wait")
```

... except it doesn't seem to work in vscode default python setup (or sublime anaconda default python setup). If you enable mypy it does detect problems in vscode, but it doesn't provide a nice drop down list. I think pycharm has Literal support, but I don't know how good it is.

It's still pretty nice to look at.
```python
set_cursor("arrow")
```

Reply:
This is how it looks in PyCharm:

![image](https://user-images.githubusercontent.com/13382426/100515349-632fa900-3173-11eb-9388-e9a0c49b221c.png)


Which I would be pretty happy with since it seems fairly close to a Enum's behaviour in terms of autocomplete in editor. Seem to be and idea borrowed from Typescript to replace Enums there which is nice. 

Apparently they are new in the typing module in 3.8 but are in the other versions of python via typing_extensions. Looks like we are already using them in `pixelcopy.pyi` as well so I would endorse this route.